core/base/v7/src/TLogger.cxx

```diff
@@ -14,32 +14,49 @@
  *************************************************************************/
 
 #include "ROOT/TLogger.hxx"
+
+#include <atomic>
 #include <iostream>
+#include <sstream>
+
+#include "TError.h"
+#include "TSpinLockGuard.h"
 
 // pin vtable
 ROOT::Experimental::TLogHandler::~TLogHandler() {}
 
 namespace {
 class TLogHandlerDefault: public ROOT::Experimental::TLogHandler {
+   std::atomic_flag fSpinLock;
+
 public:
+   TLogHandlerDefault() { std::atomic_flag_clear(&fSpinLock); }
+
    // Returns false if further emission of this log entry should be suppressed.
    bool Emit(const ROOT::Experimental::TLogEntry &entry) override;
 };
 
 bool TLogHandlerDefault::Emit(const ROOT::Experimental::TLogEntry &entry)
 {
    constexpr static std::array<const char *, 5> sTag{{"Debug", "Info", "Warning", "Log", "FATAL"}};
-   std::cerr << "ROOT ";
+   std::stringstream strm;
+   strm << "ROOT ";
    if (!entry.fGroup.empty())
-      std::cerr << '[' << entry.fGroup << "] ";
-   std::cerr << sTag[static_cast<int>(entry.fLevel)];
+      strm << '[' << entry.fGroup << "] ";
+   strm << sTag[static_cast<int>(entry.fLevel)];
 
    if (!entry.fFile.empty())
-      std::cerr << " " << entry.fFile << ':' << entry.fLine;
+      strm << " " << entry.fFile << ':' << entry.fLine;
    if (!entry.fFuncName.empty())
-      std::cerr << " in " << entry.fFuncName;
-   std::cerr << ":\n"
-             << "   " << entry.str() << '\n';
+      strm << " in " << entry.fFuncName;
+
+   {
+      ROOT::Internal::TSpinLockGuard slg(fSpinLock);
```

Comment:
Should the spin lock be here or inside the ErrorHandler?  I.e. it might be redundant with what the user does.  (And worse since we do not know what the user does there, this might lead to an easy to get-into dead-lock (for example if some ROOT code that take the ROOT global lock calls the error handler and then the user error handler (in another thread) tries to call something that require the ROOT global lock  ....

Reply:
I understand the problem, do you know of a solution?

The only one I can suggest is to stuff the lock into ROOT's error handler, and let the user- (aka framework-) provided ones deal with synchronization themselves. Makes sense?

Reply:
Yes, I agree.  We 'just' have to document that for the logger to be thread safe the user-provided call-back must be thread safe on its own.